pkg/terrors/terrors.go

```diff
@@ -0,0 +1,90 @@
+package terrors
+
+import (
+	"fmt"
+
+	"golang.org/x/xerrors"
+)
+
+// TodoneError サーバ-クライアント間エラーハンドリング用エラー
+type TodoneError struct {
+	// エラーコード
+	ErrorCode int
+	// システムエラーメッセージ(日本語)
+	ErrorMessageJP string
+	// システムエラーメッセージ(英語)
+	ErrorMessageEN string
+	// xerrors拡張用フィールド
+	err error
+	// それぞれでfmt.Errorf("%w", err)を記述する必要があるためxerrors使う。
+	frame xerrors.Frame
+}
+
+// New TodoneErrorを生成する
+func New(errorCode int) error {
+	return newError(nil, errorCode, "", "")
+}
+
+// Newf TodoneErrorをエラーメッセージ付きで生成する
+func Newf(errorCode int, messageJP string, messageEN string) error {
+	return newError(nil, errorCode, messageJP, messageEN)
+}
+
+// Wrap エラーをTodoneエラーでラップする
+func Wrap(cause error, errorCode int) error {
+	return newError(cause, errorCode, "", "")
+}
+
+// Wrapf エラーをTodoneエラーで、エラーメッセージ付きでラップする
+func Wrapf(cause error, errorCode int, messageJP, messageEN string) error {
+	return newError(cause, errorCode, messageJP, messageEN)
+}
+
+func newError(cause error, errorCode int, errorMessageJP, errorMessageEN string) error {
+	return &TodoneError{
+		ErrorCode:      errorCode,
+		ErrorMessageJP: errorMessageJP,
+		ErrorMessageEN: errorMessageEN,
+		err:            cause,
+		frame:          xerrors.Caller(2),
```

Comment:
Caller(1)だと、callerは`Wrapf`とかになってしまうのか。なるほど。

Reply:
xerrors.Frameって言うフィールドが、呼び出し元のgoroutine情報を持ってくれていたりするのだけれど、xerrors.Caller()関数に渡している数字はpcに記録する前にスキップするスタックフレームの数になっていて、さらにコードを中まで辿っていくと、runtimeパッケージ（標準パッケージ）のCallers()関数に行き着くんだけど、そこに詳しい情報が載っているんよ。

```go
// Callers fills the slice pc with the return program counters of function invocations
// on the calling goroutine's stack. The argument skip is the number of stack frames
// to skip before recording in pc, with 0 identifying the frame for Callers itself and
// 1 identifying the caller of Callers.
// It returns the number of entries written to pc.
//
// To translate these PCs into symbolic information such as function
// names and line numbers, use CallersFrames. CallersFrames accounts
// for inlined functions and adjusts the return program counters into
// call program counters. Iterating over the returned slice of PCs
// directly is discouraged, as is using FuncForPC on any of the
// returned PCs, since these cannot account for inlining or return
// program counter adjustment.
//go:noinline
func Callers(skip int, pc []uintptr) int {
	// runtime.callers uses pc.array==nil as a signal
	// to print a stack trace. Pick off 0-length pc here
	// so that we don't let a nil pc slice get to it.
	if len(pc) == 0 {
		return 0
	}
	return callers(skip, pc)
}
```

特に、
```go
// Callers fills the slice pc with the return program counters of function invocations
// on the calling goroutine's stack. The argument skip is the number of stack frames
// to skip before recording in pc, with 0 identifying the frame for Callers itself and
// 1 identifying the caller of Callers.
// It returns the number of entries written to pc.
```
の部分を読むと、「引数skipはpcに記録する前にスキップするスタックフレームの数であり、0はCallers自身のフレームを識別し、1はCallersの呼び出し元を識別します。」って書いてあって、pcに記録するスタックフレームを特定するための関数であることが分かる！

やから、`terrors.Stack()`と`terrors.Newf()` でskipしているスタックフレーム数が異なると言う説明になるかなー。